sys/net/link_layer/ieee802154/security.c

```diff
@@ -0,0 +1,463 @@
+    /* better decrypt block by block, than holding
+       a large buffer because I had stack issues */
+    uint8_t block_buffer1[IEEE802154_SEC_BLOCK_SIZE] = { 0 };
+    uint8_t block_buffer2[IEEE802154_SEC_BLOCK_SIZE] = { 0 };
+    uint8_t *c = *payload;
+    uint8_t *mac = *mic;
+    uint8_t size, offset, blk;
+    ieee802154_ccm_block_t Ai;
+    Ai.flags = IEEE802154_CCM_FLAG(0, 2);
+    Ai.nonce.frame_counter = htonl(aux->fc);
+    /* A better implementation would check if the received frame counter is
+       greater then the frame counter that has previously been received from
+       the other endpoint. This is done to protect against replay attacks.
+       But we do not store this information because we also do not have
+       a proper key store, to avoid complexity on embedded devices. */
+    Ai.nonce.security_level = security_level;
+    Ai.counter = 0;
+    memcpy(Ai.nonce.src_addr, src_address, IEEE802154_LONG_ADDRESS_LEN);
+    /* decrypt MIC */
+    if (mac_size) {
+        _ecb(ctx, block_buffer1, block_buffer2,
+             mac, (uint8_t *)&Ai, mac_size);
+    }
+    /* decrypt cipher */
+    for (offset = 0, blk = 0; offset < c_len; blk++) {
+        Ai.counter = htons(blk + 1);
+        offset += _ecb(ctx, block_buffer1, block_buffer2,
+                       &c[offset], (uint8_t *)&Ai, c_len - offset);
+    }
+    /* check MIC */
+    if (mac_size) {
+        memset(block_buffer2, 0, sizeof(block_buffer2));
+        Ai.flags = IEEE802154_CCM_FLAG(mac_size, 2);
+        Ai.counter = htons(c_len);
+        _cbc_next(ctx, block_buffer2, block_buffer1,
+                  (uint8_t *)&Ai, sizeof(Ai));
+        *((uint16_t *)block_buffer1) = htons(a_len);
+        size = _min(sizeof(block_buffer1) - sizeof(uint16_t), a_len);
+        memcpy(block_buffer1 + sizeof(uint16_t), header, size);
+        memset(block_buffer1 + sizeof(uint16_t) + size, 0,
+               sizeof(block_buffer1) - sizeof(uint16_t) - size);
+        _memxor(block_buffer1, block_buffer2, sizeof(block_buffer1));
+        memset(block_buffer2, 0, sizeof(block_buffer2));
+        ctx->dev.cipher_ops->cbc(&ctx->dev, block_buffer2, block_buffer2, block_buffer1, 1);
+        for (offset = size; offset < a_len;) {
+            offset += _cbc_next(ctx, block_buffer2,
+                                block_buffer1, &header[offset],
+                                a_len - offset);
+        }
+        for (offset = 0, blk = 0; offset < c_len; blk++) {
+            offset += _cbc_next(ctx, block_buffer2,
+                                block_buffer1, &c[offset],
+                                c_len - offset);
+        }
+        if (memcmp(block_buffer2, *mic, mac_size)) {
+            return -IEEE802154_SEC_MAC_CHECK_FAILURE;
+        }
+    }
+    return IEEE802154_SEC_OK;
+}
+
+void ieee802154_sec_set_key(ieee802154_sec_dev_t *ctx,
+                            const uint8_t *key, uint8_t key_size)
+{
+    /* NOP */
```

Comment:
What happens with the key from the `NETOPT_ENCRYPTION_KEY` set operation?

Reply:
It´s copied to `ieee802154_sec_context::cipher_t::cipher_context_t context`. The copying is done in `static void _set_key()`, which wraps around the `set_key` callback  and then copies.
For the software encryption / decryption, there is nothing else to do, hence the NOP.
For hardware support, the key must be transfered to the transceiver.

Reply:
So this is an intentional NOP, not a TODO - maybe copy that comment then :wink: